### PR TITLE
Added SpatialAdaptiveMaxPooling

### PR DESCRIPTION
**Pull request by fmassa**
https://github.com/torch/nn/issues/141

### CODE REVIEW

File: SpatialAdaptiveMaxPooling.cu [hunk 1]
```diff
@@ -0,0 +1,372 @@
+
+#define CUDA_MAX_THREADS 1024   // this is safe, in reality 256 is our limit
```
Comment (soumith) [inline SpatialAdaptiveMaxPooling.cu]: can you undef this at the end of the file.

**Comment by soumith** [comment]
just one small comment, otherwise looks good i think

File: SpatialAdaptiveMaxPooling.cu (hunk 1 above)
Comment (szagoruyko) [inline SpatialAdaptiveMaxPooling.cu]: I guess ```const int CUDA_NUM_THREADS = 1024;``` is better

Comment (soumith) [inline SpatialAdaptiveMaxPooling.cu]: hmm, if it is scoped in a function, yes. not in global scope.

Comment (szagoruyko) [inline SpatialAdaptiveMaxPooling.cu]: no it's used inside in the kernel, but it's like this in SpatialMaxPooling.cu, SpatialAveragePooling.cu and SubSampling.cu, without undef

**Comment by soumith** [comment]
torch/nn#141 had a build-break. can you please retest that stuff and send it in. i'll test it on my end and then merge (this PR as well)

**Comment by szagoruyko** [comment]
ah we forgot to commit the main file SpatialMaxPooling.lua, will fix it now sorry

**Comment by szagoruyko** [comment]
checked nn version, looks good, checking cunn

**Comment by szagoruyko** [comment]
ok both nn and cunn versions work, tested on 3 machines with unit tests, ready to be merged.

**Comment by soumith** [comment]
thanks fmassa and szagoruyko!

**Pull request merged**